text/0141-linking-traces.md

```diff
@@ -0,0 +1,447 @@
+
+Therefore, on root span start:
+  * Check if there is a previous span context stored
+    * If yes, add the span link with the `'sentry.link.type': 'previous_trace'` attribute
+  * Store the root span context as the previous root span in a storage mechanism of choice (e.g. `sessionStorage` in the browser)
+
+SDKs are free to implement heuristics around how long a previous trace span context should be considered (max time) and store additional necessary data.
+
+This means that the last-started root span will be denote the previous trace. In situations, where multiple root spans are started in parallel, the last started span "wins". However, the multiple root spans aspect of this should play a negligible role in Sentry SDKs, given our SDKs don't allow for this by default and it would require significant manual user setup. We acknowledge though that this could lead to potentially confusing relationships. 
+
+The following diagram shows optimal examples with a `tracesSampleRate` of `1.0`, meaning every span tree is positively sampled, sent to Sentry and therefore connected via a span link: 
+
+![](./0141-previous-trace-links-full-sampling.png)
+
+As one can see from the diagram, `previous_trace` span links are only necessary on frontend SDKs today. For traces started on the backend (like in Case 2), the trace links are on `pageload` transactions which are **not** the head-of-trace transactions. Our product should still be able to link the traces correctly in the trace view.
+
+### Dealing with negatively sampled traces
+
+In many cases, with lower sample rates, we will not be able to provide a full trace link chain, due to some or many traces being negatively sampled. However, sampled root spans should still include a link to the previous, negatively sampled root span. The `traceFlags` on the `spanContext()` will carry the information that the previous trace root span was negatively sampled. This helps our product to hint that there _would have been_ a previous trace but it was negatively sampled. 
+
+![](./0141-previous-trace-partial-sampling.png)
+
+It's worth noting that we will not link to the previous positively sampled trace if a negatively sampled trace is in-between (see Traces 2-4 in the diagram). Furthermore, we will not show _how many_ traces were negatively sampled in between two trace chains; only that there was at least one trace in between (see Trace 5-8). 
+
+There are some ideas how we could show multiple negatively sampled spans as well as how we could connect traces with gaps due to sampling. For now, we'll disregard this because there's no concrete use case, yet. However, given spans can have multiple links and links can have attributes, we could add this information later on, if we deem it necessary.
```

Comment:
Most customers don't sample at 100%. They will frequently experience previously negatively sampled traces. I can imagine that some customers would like to navigate to the last positively sampled trace when most of their previous traces are negatively sampled. But it might be OK to wait for customers to ask for that feature. The current proposal would allow multiple options to achieve this without changing the APIs and the protocol.

Reply:
Yup. I sketched out a solution a while ago how we could add multiple links to indicate missing traces due to sampling: 
![image](https://github.com/user-attachments/assets/5796d069-72c8-4acd-964c-80b5d5dfba29)

I didn't include this in the RFC because as I wrote, it's something we can tackle if there's demand for it.